### ISSUE TITLE
Close and Restart the Durable Functions

### ISSUE THREAD

**Issue by TsuyoshiUshio** [Azure/azure-functions-durable-extension#366]
Several people request this feature. 

Especially for Timer user and DevOps pipeline for preventing accept the new state for the new deployment.

The basic feature is to close/open Durable Functions. If we call the OrchestratorClient  StartNewAsync, then throws Specific Exception and have a state of this to the storage account. 

 If once this becomes high priority, I'd happy to contribute that.

**Comment by gled4er** [comment]
Hello @TsuyoshiUshio,

Do you think your idea is related to [#301](https://github.com/Azure/azure-functions-durable-extension/issues/301)?

There is an active pull request for that here - https://github.com/Azure/azure-functions-durable-extension/pull/341 

Thank you!

**Comment by TsuyoshiUshio** [comment]
Hi @gled4er  this is not that. 
It close to accept the new request. In case of http request trigger calls OrchestrationClient, we can implement it by ourselves. however, timer trigger can't handle that. Also, customer always implement this feature for safe deployment. It is one of request of my customer.

**Comment by jeffhollan** [comment]
If I understand correctly the desired behavior is:
1. If I "close" a durable orchestration, currently in-flight instances will continue. However new instance attempts will throw an exception
   - Potentially I could re-queue these requests, or in the case of a timer just wait for the next recurrence interval
2. At some point in the future you can "open" the orchestration (I'm guessing once in-flight have all completed) and accept new work (after a deployment for example)

**Comment by TsuyoshiUshio** [comment]
@jeffhollan  Yes. It is exactly the behavior what I want. :)

**Comment by KenKilty** [comment]
@jeffhollan / @TsuyoshiUshio There is a need for this capability.  An approach I have seen that validates your proposed approach with pre-compiled AF's is using a singleton timer job as the DF orchestration client with a app setting based enabled/disable bool value to "turn off" everything.  The approach above would remove the need for all of this could which would be great.

**Comment by SimonLuckenuik** [comment]
@jeffhollan, that would actually be useful at the Azure Functions level, not only for Durable Functions. Basically, suspending any Async Trigger (queue, blob,...) and returning a standard HTTP error code for sync trigger. Ideally, if we should be able to suspend sync and async triggers separately.

Example: I suspend async triggers 5 minutes before my deployment, but HTTP triggers are still executing 
 and can enqueue new items (but new items won't be processed right away because async triggers are suspended).

**Comment by SimonLuckenuik** [comment]
In the example from @TsuyoshiUshio, StartNewAsync would not throw, it would simply enqueue the work to be executed asynchronously and would be executed once async triggers are enabled.

**Comment by cgillum** [comment]
I'm actually seeing two slightly different proposals here:

1. **Stop new orchestration instances from starting but let existing instances continue**: This would have to be a Durable Functions-specific feature since the same async trigger processes both new and in-flight instances.
2. **Stop all async processing**: This could be implemented in the Azure Functions host and work across all trigger types. For Durable, it would pauses execution of both new and in-flight orchestrations.

Which would be more helpful? I'm personally leaning towards the second proposal because I think many scenarios might benefit from it.

**Comment by TsuyoshiUshio** [comment]
Agree, I was too much focused Durable however, "Azure Functions" might need this feature. Thank you guys for your comments! 2 sounds more helpful for me.

However, I have question for @cgillum for 2. 

You mentioned that it pauses in-flight orchestrations, if we pauses execution, how we can recover from that state?  Pause -> Restart scenario is ok but, I want to for Graceful shutdown and new app deployment.  Then just pause the execution could cause a problem (Pause -> Deploy new version -> Accept new request again) As @SimonLuckenuik  if we've got new request, enqueue the work, leave the in-flight instance after becoming acceptance new request state, it consume the queue or just through an error (it might be depends on trigger or usecase)   Am I understand what you said? 

Also, if we choose 2. option, I guess only Durable might have a problem (other triggers are ok), If I want to have Graceful shutdown for new deployment, I turn the flag to "Stop accepting the new request"  
then , Orchestration/ActivityTrigger stop accepting the request. Which means if we have in-flight instances, the in-flight instances never complete. For Durable, once stop accepting, leave the in-flight instances until all of these completed might be good for this scenario.

So maybe we need to implement both Azure Functions and Durable Functions. For Azure Functions, in-flight instances mean Single Function execution. But For Durable, it means orchestration (with several Single Function execution instances)

For example, once we change the AppSettings (e.g. STOP_ACCEPTING_NEW_REQUEST=true), it stop accepting new request for Azure Functions wide, (however, it could cause problem for Durable), for case of Durable, Durable need to have some flag inside of Storage account, check when it accept a new request.

**Comment by TsuyoshiUshio** [comment]
I was wrong. Changing AppSettings could lead restarting. Does ScriptJobHost have some APIs like Durable?

**Comment by SimonLuckenuik** [comment]
@cgillum, I prefer "2. Stop all async processing". Just to clarify my understanding, it would:
1) stop processing incoming any async triggers (queues, blob,...)
2) let any executing functions complete normally
3) let any orchestrator complete normally, until reaching next activity, next activity is submitted, but not executed since we are not processing any incoming async triggers

**Comment by codeputer** [comment]
I can't tell you how many times I've gone back to a function, and threads and breakpoints are hit out of order while in debug mode due to multi-threads starting during a durable function.  I currently simply clear the emulator storage, which allows me to "reset" durable functions.

**Comment by YunchuWang** [comment]
Thanks for raising this. Closing this issue because the original safe-deployment scenario has been superseded by newer platform and Durable Functions capabilities.

The recommended solution is now [orchestration versioning](https://learn.microsoft.com/azure/durable-task/common/durable-orchestration-versioning). It permanently associates each orchestration instance with a version and allows compatible old and new orchestration code to coexist during rolling or blue/green deployments, without first draining or abandoning all in-flight orchestration instances. The Azure Functions platform also drains currently executing function invocations before shutting down a host instance; this is invocation draining, not a guarantee that an entire long-running orchestration reaches a terminal state before deployment.

This does **not** mean that the literal `Close`/`Open` admission-control API proposed here was added. Remaining scenarios that require explicit coordinated shutdown—for example, rejecting new starts or activities while allowing existing work to drain—continue to be tracked in #658.

Versioning initially landed for .NET in-process in #3072, then expanded through #3095, #3137, and #3213. It is now documented for .NET isolated, JavaScript, Python, PowerShell, and Java, so the replacement is not limited to .NET in-process.
